node/compile-hook.js

```diff
@@ -11,10 +12,31 @@ if (! _compile.reified) {
   (Mp._compile = function (content, filename) {
     return _compile.call(
       this,
-      // Don't touch the file unless reification is enabled for the
-      // package that contains the file.
-      compile(content, filename),
+      compile(content, getOptions(filename)),
       filename
     );
   }).reified = _compile;
 }
+
+function getOptions(filename) {
+  const options = { filename: filename };
+  const mtime = mtimeOrNull(filename);
+
+  if (mtime !== null) {
+    options.cacheKey = {
+      source: "Module.prototype._compile",
+      filename: filename,
+      mtime: mtime
+    };
```

Comment:
JSON stringifying a date will produce a string.
```js
JSON.stringify({ 'a': new Date })
// => '{"a":"2017-03-28T21:36:01.892Z"}'
```
You could have it be a timestamp which I think is a bit better
```js
mtime: mtime.getTime()
```